### PR TITLE
snap/naming: move various name validation helpers to separate package

### PR DESCRIPTION
**Pull request by bboozzoo**
Follow-up on the package separation idea discussed in #5948. The PR moves various snap related name validation helpers into a separate package that can be easily imported without pulling in the whole of `githut.com/snapcore/snapd/snap` baggage.

As part of the PR, the package is used by model assertion validation code, where the `snap` package could not have been previously used.

### CODE REVIEW

**Comment by chipaca** [comment]
wouldn't `snap/validate` with e.g. `validate.Plug()` functions, be nicer to use?

**Comment by bboozzoo** [comment]
@chipaca that's something I thought of in the beginning. The downside is, you'd still need to have access to `snap.Info` to move the whole of validation code into a separate package, so `validate` would need to import `snap` or some other package that brings `Info` (snap/meta?). On the other hand, I'm really only interested in having access to `snap.ValidateName()` and maybe `ValidateInstanceName()` from `asserts`. I'm open to suggestions though. Maybe we could move the 'heavy' code out of `snap`?

**Comment by mvo5** [comment]
I think this is nice - I am not sure if I understand @chipaca correct but I also prefer `snap/validate' over `snap/name` or `snap/validatenames` (but thats a bit heavy). I mean, right now it seems all the exported functions are Validate{Instance,Snap,Plug,Slot,Hook,Alias,App,Socket}` so why not `validate.SnapName()` or something along these lines?

**Comment by zyga** [comment]
Please deconflict @bboozzoo

**Comment by bboozzoo** [comment]
As discussed, I've renamed the package to `snap/naming`. The changed code reads surprisingly nice.

File: snap/validate.go [hunk 1]
```diff
@@ -30,89 +30,32 @@ import (
...
-	storeName := instanceName[:pos]
-	instanceKey := instanceName[pos+1:]
-	if err := ValidateName(storeName); err != nil {
-		return err
-	}
-	if !validInstanceKey.MatchString(instanceKey) {
-		return fmt.Errorf("invalid instance key: %q", instanceKey)
-	}
-	return nil
+	return naming.ValidateInstance(instanceName)
```
Comment (pedronis) [inline snap/validate.go]: what are you thoughts? is that a intermediate step for churn? I would suppose in the end we don't want this indirection functions

Comment (bboozzoo) [reply snap/validate.go]: Yes, I didn't want the diff to be overly large. The cleanup can be introduced once this PR lands.

File: snap/naming/validate.go [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+	}
+
+	storeName := instanceName[:pos]
+	instanceKey := instanceName[pos+1:]
+	if err := ValidateSnap(storeName); err != nil {
+		return err
+	}
+	if !validInstanceKey.MatchString(instanceKey) {
+		// TODO parallel-install: extend the error message once snap
+		// install help has been updated
```
Comment (pedronis) [inline snap/naming/validate.go]: is this TODO still relevant? something we can do now?

File: snap/naming/validate.go [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+//   regexp package doesn't support look-aheads nor look-behinds, so in order to
+//   have a unified implementation in the Go and Python bits of the project
+//   we're doing it this way instead. Check the length (if applicable), check
+//   this regexp, then check the dashes.
+//   This still leaves sc_snap_name_validate (in cmd/snap-confine/snap.c) and
+//   snap_validate (cmd/snap-update-ns/bootstrap.c) with their own handcrafted
+//   validators.
+var almostValidName = regexp.MustCompile("^[a-z0-9-]*[a-z][a-z0-9-]*$")
+
+// validInstanceKey is a regular expression describing valid snap instance key
```
Comment (pedronis) [inline snap/naming/validate.go]: a valid

File: snap/naming/validate.go [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package naming
```
Comment (pedronis) [inline snap/naming/validate.go]: // Package naming implements naming constraints for snaps and their elements.

File: snap/naming/validate.go [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+//   this regexp, then check the dashes.
+//   This still leaves sc_snap_name_validate (in cmd/snap-confine/snap.c) and
+//   snap_validate (cmd/snap-update-ns/bootstrap.c) with their own handcrafted
+//   validators.
+var almostValidName = regexp.MustCompile("^[a-z0-9-]*[a-z][a-z0-9-]*$")
+
+// validInstanceKey is a regular expression describing valid snap instance key
+var validInstanceKey = regexp.MustCompile("^[a-z0-9]{1,10}$")
+
+// isValidName checks snap and socket socket identifiers.
```
Comment (pedronis) [inline snap/naming/validate.go]: s/socket socket/socket/ ?

**Pull request merged**